src/coreComponents/linearAlgebra/interfaces/trilinos/TrilinosSolver.cpp

```diff
@@ -54,71 +57,141 @@ void TrilinosSolver::solve( EpetraMatrix & mat,
 
   GEOSX_UNUSED_VAR( dofManager );
 
-  if( m_parameters.scaling.useRowScaling )
+  if( rhs.norm2() > 0.0 )
   {
-    Epetra_FECrsMatrix & mat_raw = mat.unwrapped();
-    Epetra_MultiVector & rhs_raw = rhs.unwrapped();
+    if( m_parameters.scaling.useRowScaling )
+    {
+      Epetra_FECrsMatrix & mat_raw = mat.unwrapped();
+      Epetra_MultiVector & rhs_raw = rhs.unwrapped();
 
-    Epetra_Vector scaling( mat_raw.RowMap() );
-    mat_raw.InvRowSums( scaling );
-    mat_raw.LeftScale( scaling );
+      Epetra_Vector scaling( mat_raw.RowMap() );
+      mat_raw.InvRowSums( scaling );
+      mat_raw.LeftScale( scaling );
 
-    Epetra_MultiVector tmp( rhs_raw );
-    rhs_raw.Multiply( 1.0, scaling, tmp, 0.0 );
-  }
+      Epetra_MultiVector tmp( rhs_raw );
+      rhs_raw.Multiply( 1.0, scaling, tmp, 0.0 );
+    }
 
-  if( m_parameters.solverType == LinearSolverParameters::SolverType::direct )
-  {
-    solve_direct( mat, sol, rhs );
+    if( m_parameters.solverType == LinearSolverParameters::SolverType::direct )
+    {
+      solve_direct( mat, sol, rhs );
+    }
+    else
+    {
+      solve_krylov( mat, sol, rhs );
+    }
   }
   else
   {
-    solve_krylov( mat, sol, rhs );
+    sol.zero();
+    m_result.status = LinearSolverResult::Status::Success;
+    m_result.setupTime = 0.0;
+    m_result.solveTime = 0.0;
   }
 }
 
-void TrilinosSolver::solve_direct( EpetraMatrix & mat,
-                                   EpetraVector & sol,
-                                   EpetraVector & rhs )
+namespace
 {
-  // Time setup and solve
-  Stopwatch watch;
 
-  // Create Epetra linear problem and instantiate solver.
-  Epetra_LinearProblem problem( &mat.unwrapped(),
-                                &sol.unwrapped(),
-                                &rhs.unwrapped() );
+void solve_parallelDirect( LinearSolverParameters const & parameters,
+                           EpetraMatrix & mat,
+                           EpetraVector & sol,
+                           EpetraVector & rhs,
+                           LinearSolverResult & result )
+{
+  // To be able to use SuperLU_Dist solver we need to disable floating point exceptions
+  LvArray::system::FloatingPointExceptionGuard guard;
 
-  // Instantiate the Amesos solver.
-  Amesos Factory;
+  SuperLU_Dist SLUDData( parameters );
+  EpetraConvertToSuperMatrix( mat, SLUDData );
 
-  // Select KLU solver (only one available as of 9/20/2018)
-  std::unique_ptr< Amesos_BaseSolver > solver( Factory.Create( "Klu", problem ) );
+  GEOSX_LAI_CHECK_ERROR( SLUDData.setup() );
+  GEOSX_LAI_CHECK_ERROR( SLUDData.solve( rhs.extractLocalVector(), sol.extractLocalVector() ) );
```

Comment:
If Trilinos were built with SuperLU-Dist ideally this would be simpler and `EpetraConvertToSuperMatrix` wouldn't be necessary.